ext/opencensus_trace.c

```diff
@@ -657,22 +664,46 @@ void opencensus_trace_execute_internal(INTERNAL_FUNCTION_PARAMETERS)
     );
     zval *trace_handler;
     opencensus_trace_span_t *span;
+    zend_string *callback_name = NULL;
 
-    if (function_name) {
-        trace_handler = zend_hash_find(OPENCENSUS_TRACE_G(user_traced_functions), function_name);
+    /* Some functions have no names - just execute them */
+    if (function_name == NULL) {
+        resume_execute_internal(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+        return;
+    }
 
-        if (trace_handler) {
-            span = opencensus_trace_begin(function_name, execute_data, NULL TSRMLS_CC);
-            resume_execute_internal(INTERNAL_FUNCTION_PARAM_PASSTHRU);
-            opencensus_trace_execute_callback(span, execute_data, trace_handler TSRMLS_CC);
-            opencensus_trace_finish();
-        } else {
-            resume_execute_internal(INTERNAL_FUNCTION_PARAM_PASSTHRU);
-        }
+    trace_handler = zend_hash_find(OPENCENSUS_TRACE_G(user_traced_functions), function_name);
+
+    /* Function is not registered for execution - continue normal execution */
+    if (trace_handler == NULL) {
+        resume_execute_internal(INTERNAL_FUNCTION_PARAM_PASSTHRU);
         zend_string_release(function_name);
+        return;
+    }
+
+    span = opencensus_trace_begin(function_name, execute_data, NULL TSRMLS_CC);
+    zend_string_release(function_name);
+
+    if (zend_is_callable(trace_handler, 0, &callback_name)) {
+        /* Registered handler is callable - execute the callback */
+        zval callback_result, *args;
+        int num_args;
+        opencensus_copy_args(execute_data, &args, &num_args);
+        resume_execute_internal(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+        if (opencensus_trace_call_user_function_callback(args, num_args, execute_data, span, trace_handler, &callback_result TSRMLS_CC) == SUCCESS) {
+            opencensus_trace_span_apply_span_options(span, &callback_result);
+        }
```

Comment:
Is it worth emitting a warning when the user callback fails here?

Reply:
We already raise a warning inside `opencensus_trace_call_user_function_callback` if there's an exception inside of the callback or if the return value is not an array of span attributes. Any other failure is a PHP internal error that should already raise a warning. It's possible that we also raise a warning with a generic message for these cases as well.

Reply:
LGTM